File: 95.c
#include <stdio.h>

#define SIZE 1000000

int sdiv[SIZE];
int hsh[SIZE];

int chain(int n) {
	int i;
	for (i=0; i < SIZE; i++) { hsh[i] = 0; }
	hsh[n] = 1;
	int cur = sdiv[n];
	int min = n;
	int cnt = 1;
	while(cur < SIZE && hsh[cur] != 1) {
		cnt++;
		hsh[cur] = 1;
    if (cur < min) {
			min = cur;
    }
		cur = sdiv[cur];
	} 
  if (cur == n) {
		return(cnt);
  } else {
		return(-1);
  }
}


int main() {
	int i,d,cur;
	int best = 0;
	for (i=1; i < SIZE; i++) {
		sdiv[i] = 0;
	}
	for (d=1; d < SIZE; d++) {
		i = d * 2;
		while (i < SIZE) {
			sdiv[i]+= d;
			i += d;
		}
	}
	for (i=1; i < SIZE ; i++) {
		cur = chain(i);
	}
	printf("%d\n", chain(12496));
	return(0);
	
}

File: 95.rb
require 'pp'

SIZE = 1_000_000

@sdiv = Array.new(SIZE,0)

1.upto(SIZE) do |d|
  i = d*2
  while i < SIZE
    @sdiv[i]+= d
    i+= d
  end
end


@max_length = 0
@min = 0
1.upto(SIZE-1) do |i|
  h = Hash.new(false)
  j = i
  while !h[j]
    h[j] = true
    j = @sdiv[j]
    break if j > SIZE-1
  end
  if j == i && h.keys.size > @max_length
    @max_length = h.keys.size
    @min = h.keys.min
  end
end

pp @max_length
gpp @min
